tests/templates/test_state_preparations/test_arbitrary_state_prep.py

```diff
@@ -0,0 +1,351 @@
+            qml.templates.ArbitraryStatePreparation(weights, [0, 1, 2])
+
+            return qml.expval(qml.PauliZ(0))
+
+        circuit(weights)
+
+        assert np.allclose(circuit.device.state, even_superposition_state, atol=tol, rtol=0)
+
+    def test_custom_wire_labels(self, tol):
+        """Test that template can deal with non-numeric, nonconsecutive wire labels."""
+        weights = np.random.random(size=(2 ** 4 - 2))
+
+        dev = qml.device("default.qubit", wires=3)
+        dev2 = qml.device("default.qubit", wires=["z", "a", "k"])
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.templates.ArbitraryStatePreparation(weights, wires=range(3))
+            return qml.expval(qml.Identity(0))
+
+        @qml.qnode(dev2)
+        def circuit2():
+            qml.templates.ArbitraryStatePreparation(weights, wires=["z", "a", "k"])
+            return qml.expval(qml.Identity("z"))
+
+        circuit()
+        circuit2()
+
+        assert np.allclose(dev.state, dev2.state, atol=tol, rtol=0)
+
+
+class TestInputs:
+    """Test inputs and pre-processing."""
+
+    def test_exception_wrong_dim(self):
+        """Verifies that exception is raised if the
+        number of dimensions of features is incorrect."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def circuit(weights):
+            qml.templates.ArbitraryStatePreparation(weights, wires=range(3))
+            return qml.expval(qml.PauliZ(0))
+
+        with pytest.raises(ValueError, match="Weights tensor must be of shape"):
+            weights = np.zeros(12)
+            circuit(weights)
+
+
+class TestAttributes:
+    """Tests additional methods and attributes"""
+
+    @pytest.mark.parametrize(
+        "n_wires, expected_shape",
+        [
+            (3, (14,)),
+            (1, (2,)),
+            (2, (6,)),
+        ],
+    )
+    def test_shape(self, n_wires, expected_shape):
+        """Test that the shape method returns the correct shape of the weights tensor"""
+
+        shape = qml.templates.ArbitraryStatePreparation.shape(n_wires)
+        assert shape == expected_shape
+
+
+def circuit_template(weights):
+    qml.templates.ArbitraryStatePreparation(weights, range(2))
+    return qml.expval(qml.PauliZ(0))
+
+
+def circuit_decomposed(weights):
+    qml.PauliRot(weights[0], "XI", wires=[0, 1])
+    qml.PauliRot(weights[1], "YI", wires=[0, 1])
+    qml.PauliRot(weights[2], "IX", wires=[0, 1])
+    qml.PauliRot(weights[3], "IY", wires=[0, 1])
+    qml.PauliRot(weights[4], "XX", wires=[0, 1])
+    qml.PauliRot(weights[5], "XY", wires=[0, 1])
+    return qml.expval(qml.PauliZ(0))
+
+
+class TestInterfaces:
+    """Tests that the template is compatible with all interfaces, including the computation
+    of gradients."""
+
+    def test_list_and_tuples(self, tol):
+        """Tests common iterables as inputs."""
+
+        weights = [1] * 6
+
+        dev = qml.device("default.qubit", wires=2)
```

Comment:
Maybe instead take `qubit_device_2_wires` as a fixture input?

Reply:
I think our current idea is to move away from fixtures, to make the tests as self-contained as possible. Ideally we should be able just to copy-and-paste the test and it runs!

Reply:
Good to know